### PR TITLE
Correct topological sorting of declarations.

### PR DESCRIPTION
**Pull request by xuhdev**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29238 Remove specific processing of declarations of add, sub, mul, div, rsub
* **#29237 Correct topological sorting of declarations.**

The current algorithm implements Kahn's algorithm, but it seems to have put the order completely reversed. For example, in `python_torch_functions.py` currently generates

```python
  static PythonArgParser parser({
    "pow(Tensor input, Tensor exponent, *, Tensor out=None)",
    "pow(Scalar self, Tensor exponent, *, Tensor out=None)",
    "pow(Tensor input, Scalar exponent, *, Tensor out=None)",
  }, /*traceable=*/true);
```

The order should be completely reversed so that the scalar version can be chosen first. This is also true when exposing the scalar version of add, sub, mul, div, rsub to Python in #29238. With the current ordering, the scalar version is never used.

### CODE REVIEW

**Comment by gchanan** [comment]
description?  What's the problem?

**Comment by xuhdev** [comment]
@gchanan Updated

**Comment by VitalyFedyunin** [comment]
Please cleanup test failures and re-add me as reviewer.

File: tools/autograd/gen_python_functions.py [hunk 1]
```diff
@@ -881,7 +881,7 @@ def is_smaller(d1, d2):
                 del larger_than[i2]
                 sorted_deps.append((i2, grouped_decls[i2]))
 
-    return [decl for i, decl in sorted_deps]
+    return [decl for i, decl in reversed(sorted_deps)]
```
Comment (gchanan) [inline tools/autograd/gen_python_functions.py]: comment.

**Review by gchanan — CHANGES_REQUESTED** [review]
I'm a bit confused how this interacts with the stacked PR.

What's the state of world when this goes in but the next one hasn't yet?  (Actually, all three cases would be good to understand, i.e.):
1) now
2) after this
3) after the stack

**Comment by xuhdev** [comment]
> I'm a bit confused how this interacts with the stacked PR.

After a long time investigation, I have to admit that I probably have mistaken. I write my original thought below, just in case I double mistook myself (Since much of the following is no longer true, I'm a bit confused by the purpose of the changed topological sort):

The stacked PR won't work properly without ordering declarations correctly, because, based on the current ordering, enabling the Scalar version of add/sub/... only adds them to the end of the list of declarations, and thus they are still not called because the tensor version would match first (**this seems not true any more, at least for the current master**). For example, after this PR, the stacked PR would generate in `torch/csrc/autograd/generated/python_torch_functions.cpp`:

```cpp
   static PythonArgParser parser({
    "add(Tensor input, Scalar other, Scalar alpha=1)",
     "add(Tensor input, Scalar alpha, Tensor other, *, Tensor out=None)|deprecated",
     "add(Tensor input, Tensor other, *, Scalar alpha=1, Tensor out=None)",
   }, /*traceable=*/true);
```

Before this PR, the stacked PR would generate

```cpp
   static PythonArgParser parser({
     "add(Tensor input, Scalar alpha, Tensor other, *, Tensor out=None)|deprecated",
     "add(Tensor input, Tensor other, *, Scalar alpha=1, Tensor out=None)",
     "add(Tensor input, Scalar other, Scalar alpha=1)",
   }, /*traceable=*/true);
```

In this order, when `torch.add(torch.tensor([1, 1]), 2)` is called, the second declaration would take effect, instead of the third one (**not true on current master**). I believe this is also the reason we have the topological sort at issue here, according to a comment above the changed lines:

```python
# This function declares a partial order on declarations, and sorts them according
# to its linear extension. This is necessary, because there's some ambiguity in the
# choice of overload, and we want a different order.
```

**Comment by xuhdev** [comment]
I'm closing this now so as to have a fresh CI run for the stacked PR

**Pull request closed without merging**